proposal.rst

```diff
@@ -166,27 +162,41 @@ Project Details
 
 **Describe any existing work and concepts on which your project is based.**
 
-FILLME
+The project is divided into 4 subprojects -
+
+1) Many modern file formats are based around the zip file format with XML contents. So using a zip file reading library instead of the unzip program would is the first target. Currently, Omega uses zlib to read gzip compressed Abiword files. To cover these formats - using libarchive is probably a sensible option. Implementing this library is the first subproject.
+2) Currently the extraction of text from a PDF file is done by running external commands pdfinfo and pdftotext on it and reading their output whereas those two commands use libpoppler to do their work. Hence, using libpoppler and saving running two commands per PDF file would be desirable instead of creating two child processes, and two times any set up and tear down libpoppler doe. Implementing this library is the second subproject.
+3) DjVu is a web-centric format which can display documents and images. DjVuLibre can be used as an extractor for these type of files. Implementing this library is the third subproject.
+4) There are number of other file formats which require external filter programs and can be replaced by available libraries. Although if we could not cover all the libraries available, we could give priority to the file formats which are used more than others such that the overall indexing speed can be reduced by a noticeable amount. This is the fourth subproject.
+
+The approach for making the libraries work on replacement of external filter programs with shared libraries having same functionalities and hence reducing the time required for indexing is to dynamically load at the runtime. Also, to avoid library bugs from crashing omindex, they can be implemented in a subprocess isolated from the parent process. The output of the child subprocess created using fork() syscall could be piped to the parent process. Another issue is if the library ends up in a memory or CPU eating infinite loop. This can be avoided by using sandboxing to put resource constraints on the process so that the loop will eventually terminate.
```

Comment:
It feels to me like there is either an initial subproject, or a dependent step of all of the above, to figure out some of the details of how this would work, and that you should put your current thoughts down here. (It's possible some of this is captured elsewhere, but I think it should be in your proposal.) For instance:

 * I assume you'll load the library needed in the subprocess?
 * Are subprocesses the right model on all operating systems? (Or perhaps on some we won't support this approach at all, at least initially?)
 * are you proposing to dynamically link all these libraries? Or to runtime load them?
 * are there portability concerns? (eg: linking or even compiling the libraries on Windows; efficient transfer of data between the child and parent process on different operating systems)
 * are there automated tests you can write to give confidence in how this system will work? (For instance, you could write a minimal library that can exhibit different failures depending on the file it's run over.)

Reply:
- Yes, the library will be opened using dlopen() and the functions can be loaded using dlsym() after that. 
- What I found out is that Windows doesn't support fork() and CreateProcess() is the closest thing Windows has to fork(), but Windows requires you to specify an executable to run in that process.
The UNIX process creation is quite different to Windows. Its fork() call basically duplicates the current process almost in total, each in their own address space, and continues running them separately which tells us that Windows doesn't support this approach. 
- We are using dlopen() which runtime loads the libraries
- Portability issues can arise for non *nix systems. I would need to study about that. 
- I can surely write a library and run over files to test it but I don't have the knowledge about how the code in the library extracts it. But a minimal library can be written. I'm not sure how much time it would require (firstly, I would need to read about it and write the code).

Reply:
You should include some notes in your final proposal on where you think the project will go based on what you know about this. It's okay to not have complete information! But note what you do know, and include time to read up & plan what you do not.

Reply:
I understand how to use `dlopen()` and `dlsym()`, but I'm not really following how your approach would actually work in practice - perhaps you could sketch out an example of what this looks like using a particular library?

Also could you explain why have you picked this approach over the approach used in the prototype patch linked from the project idea?

Reply:
I didn't really think of other approaches before picking up. Perhaps, I picked it because it was suggested in the idea description (it was given that use dlopen() or similar) and I saw (tried to understand) the patch well after it. So at the beginning I thought dlopen() can be used. So I read about dlopen() and I thought this may work and carried that approach forward. Can you tell me what are the problems with this approach?

Reply:
The part about `dlopen()` is talking about any libraries which we use **directly** from `omindex` itself - for example, if we want to use `libarchive` to read ZIP files then it would be quite awkward to do that work in a subprocess so we'd probably really have to use libarchive from the parent process, i.e. omindex itself.  We could make use of libarchive optional, but that would mean someone packaging omega would have to decide whether to enable it or not on behalf of all their users.  They'll probably have to enable it as some users will want it, and then all users have to install libarchive even if they don't use it.

If we use dlopen() and dlsym() for libarchive, that allows a packager to have the best of both worlds - they can enable libarchive support, but make the dependency on libarchive optional, and if it's not install omindex would still work.

But as the project idea says, this may not be worth the effort.  It really depends if the library in question is very likely to be installed anyway.  For example, we just hard depend on zlib because so many things depends on it that it's almost certain to be installed anyway.

If we're isolating the code to call libpoppler in a

Reply:
I don't really see this approach working well for something like libpoppler which we are wanting to wrap in a subprocess.

I think you must be picturing some sort of generic code which takes a library name and some function names and calls dlopen() for the library name and then dlsym() for the function names and then calls those functions to extract text from a document.

But the problem is that the APIs for different extraction libraries vary (hugely in some cases), so it seems this would end up having to create some sort of mini-scripting language in which you'd end up specifying the function call sequence to extract text using libpoppler, etc.  And that's inevitably going to be less flexible that just writing some C or C++ code to call the extraction library's API would be.

Also, while most platforms support something equivalent to `dlopen()` and `dlsym()`, the actual API varies - that means more work to support this for all the platforms omindex currently works on.  And we can't fall back to just linking the library at build time, which we could for libarchive in omindex.

The approach taken in the existing patch is that there's a class framework which takes care of running the communications between the parent and subprocess, and each extraction library is wrapped in a small program which uses the client classes from that framework and links to the extraction library (typically dynamically but it could be static and still work).

This works great for packagers - they can just put each of those small programs into its own package (and only that package needs to depend on the extraction library) and the user can choose which they want to install.

And it'll probably even work with the `CreateProcess()` API you mention since it just requires being able to run a program in a subprocess.